From 3ee1ef830acf922b4cf17fd453481b7f88fef1dc Mon Sep 17 00:00:00 2001
From: Sriram Dash <sriram.dash@nxp.com>
Date: Mon, 21 Mar 2016 16:31:07 +0530
Subject: [PATCH 39/74] drivers:usb:xhci:fsl: Change burst beat and outstanding
 pipelined transfers requests

This is required for better performance, and performs below tuning:
1. Enable burst length set, and define it as 4/8/16.
2. Set burst request limit to 16 requests.

Signed-off-by: Rajesh Bhagat <rajesh.bhagat@nxp.com>
Signed-off-by: Sriram Dash <sriram.dash@nxp.com>
---
 drivers/usb/host/xhci-fsl.c  | 12 ++++++++++++
 include/linux/usb/xhci-fsl.h |  3 +++
 2 files changed, 15 insertions(+)

diff --git a/drivers/usb/host/xhci-fsl.c b/drivers/usb/host/xhci-fsl.c
index 013e872..b5a0ae1 100644
--- a/drivers/usb/host/xhci-fsl.c
+++ b/drivers/usb/host/xhci-fsl.c
@@ -27,6 +27,15 @@ __weak int __board_usb_init(int index, enum usb_init_type init)
 	return 0;
 }
 
+static void fsl_xhci_set_beat_burst_length(struct dwc3 *dwc3_reg)
+{
+	int val = readl(&dwc3_reg->g_sbuscfg0);
+	val &= ~USB3_ENABLE_BEAT_BURST_MASK;
+	writel(val | USB3_ENABLE_BEAT_BURST, &dwc3_reg->g_sbuscfg0);
+	val = readl(&dwc3_reg->g_sbuscfg1);
+	writel(val | USB3_SET_BEAT_BURST_LIMIT, &dwc3_reg->g_sbuscfg1);
+}
+
 static int fsl_xhci_core_init(struct fsl_xhci *fsl_xhci)
 {
 	int ret = 0;
@@ -50,6 +59,9 @@ static int fsl_xhci_core_init(struct fsl_xhci *fsl_xhci)
 	dwc3_set_rxdetect_power_mode(fsl_xhci->dwc3_reg,
 				     DWC3_GUSB3PIPECTL_DISRXDETP3);
 
+	/* Change beat burst and outstanding pipelined transfers requests */
+	fsl_xhci_set_beat_burst_length(fsl_xhci->dwc3_reg);
+
 	return ret;
 }
 
diff --git a/include/linux/usb/xhci-fsl.h b/include/linux/usb/xhci-fsl.h
index c5e42e6..b028276 100644
--- a/include/linux/usb/xhci-fsl.h
+++ b/include/linux/usb/xhci-fsl.h
@@ -20,6 +20,9 @@
 #define USB3_PHY_TX_RX_POWERON	(USB3_PHY_RX_POWERON | USB3_PHY_TX_POWERON)
 #define USB3_PWRCTL_CLK_CMD_SHIFT   14
 #define USB3_PWRCTL_CLK_FREQ_SHIFT	22
+#define USB3_ENABLE_BEAT_BURST		0xF
+#define USB3_ENABLE_BEAT_BURST_MASK	0xFF
+#define USB3_SET_BEAT_BURST_LIMIT	0xF00
 
 /* USBOTGSS_WRAPPER definitions */
 #define USBOTGSS_WRAPRESET	BIT(17)
-- 
2.1.0.27.g96db324

